### ISSUE TITLE
Invalid pattern matching crashes Dart compiler

### ISSUE THREAD

**Issue by sma** [dart-lang/sdk#53753]
This hopefully minimal code example crashes the compiler of dart `3.3.0-16.0.dev`:
```dart
((int, int), int) foo() => ((1, 1), 2);

void main() {
  var b = 1;
  (final (a1, a2), b) = foo();
}
```

Here's the error:
```
$ dart run bin/a.dart 
Crash when compiling file:///Users/sma/Work/crash/bin/a.dart at character offset 51:
No name provided

#0      _withArgumentsPatternAssignmentDeclaresVariable (package:_fe_analyzer_shared/src/messages/codes_generated.dart:12145:21)
#1      Parser.parseVariablePattern (package:_fe_analyzer_shared/src/parser/parser_impl.dart:9768:33)
#2      Parser.parsePrimaryPattern (package:_fe_analyzer_shared/src/parser/parser_impl.dart:9576:16)
#3      Parser.parsePattern (package:_fe_analyzer_shared/src/parser/parser_impl.dart:9451:13)
#4      Parser.parseParenthesizedPatternOrRecordPattern (package:_fe_analyzer_shared/src/parser/parser_impl.dart:10002:15)
#5      Parser.parsePrimaryPattern (package:_fe_analyzer_shared/src/parser/parser_impl.dart:9605:15)
#6      Parser.parsePattern (package:_fe_analyzer_shared/src/parser/parser_impl.dart:9451:13)
#7      Parser.parsePatternAssignment (package:_fe_analyzer_shared/src/parser/parser_impl.dart:10185:13)
#8      Parser.parseExpression (package:_fe_analyzer_shared/src/parser/parser_impl.dart:5720:17)
#9      Parser.parseExpressionStatement (package:_fe_analyzer_shared/src/parser/parser_impl.dart:5685:13)
#10     Parser.parseExpressionStatementOrDeclarationAfterModifiers (package:_fe_analyzer_shared/src/parser/parser_impl.dart:8056:16)
#11     Parser.parseExpressionStatementOrDeclaration (package:_fe_analyzer_shared/src/parser/parser_impl.dart:7901:12)
#12     Parser.parseStatementX (package:_fe_analyzer_shared/src/parser/parser_impl.dart:5581:14)
#13     Parser.parseStatement (package:_fe_analyzer_shared/src/parser/parser_impl.dart:5476:20)
#14     Parser.parseFunctionBody (package:_fe_analyzer_shared/src/parser/parser_impl.dart:5381:15)
#15     DietListener.buildFunctionBody (package:front_end/src/fasta/source/diet_listener.dart:1232:14)
#16     DietListener.endTopLevelMethod (package:front_end/src/fasta/source/diet_listener.dart:364:5)
#17     Parser.parseTopLevelMethod (package:_fe_analyzer_shared/src/parser/parser_impl.dart:3809:14)
#18     Parser.parseTopLevelMemberImpl (package:_fe_analyzer_shared/src/parser/parser_impl.dart:3553:14)
#19     Parser.parseTopLevelDeclarationImpl (package:_fe_analyzer_shared/src/parser/parser_impl.dart:616:14)
#20     Parser.parseUnit (package:_fe_analyzer_shared/src/parser/parser_impl.dart:411:15)
#21     SourceLoader.buildBody (package:front_end/src/fasta/source/source_loader.dart:1244:12)
```

(I was hoping that I could write this in just one line:
```dart
var b = 1;
final int a1, a2;
((a1, a2), b) = foo();
```
)

**Comment by stereotype441** [comment]
Fixed by 6fcd15c1aa024bd42056487374a146be492277a2.

**Comment by stereotype441** [comment]
Fixed by 6fcd15c1aa024bd42056487374a146be492277a2.
